### PR TITLE
feat: support test source dirs in the plugin

### PR DESCRIPTION
**Pull request by MartinWitt**
Fixes https://github.com/INRIA/spoon/issues/3327

should do the trick, with adding a new parameter for test src. Sadly I need to fix maven issues before i can test it, because a dependency cant be resolved. If somebody wants to test it, it could be faster.

For the record the following fails:

```xml
<dependency>
      <groupId>fr.inria.gforge.spoon</groupId>
      <artifactId>processors</artifactId>
      <version>1.0-SNAPSHOT</version>
      <scope>test</scope>
  </dependency>
```

### CODE REVIEW

**Comment by MartinWitt** [comment]
should be working now.

File: src/test/java/fr/inria/gforge/spoon/mojo/SpoonMojoTest.java [hunk 1]
```diff
@@ -252,4 +253,39 @@ public void testSpoonGoalGenerateResultFileForProjectWithGeneratedSources() thro
...
+		rule.executeMojo(basedir, "check");
+
+		final File dirOutputResults = new File(basedir, "target/spoon-maven-plugin");
+		assertThat(dirOutputResults).exists();
+
+		final File contentSource = new File(basedir, "target/generated-sources/spoon/fr/inria/gforge/spoon");
+		assertThat(contentSource).doesNotExist();
+		
+		final File resultFileProcessor = new File(basedir, "target/spoon-maven-plugin/spoon-nb-statement.txt");
+		assertThat(resultFileProcessor).exists();
```
Comment (vmassol) [inline src/test/java/fr/inria/gforge/spoon/mojo/SpoonMojoTest.java:289]: I think this test will always pass. Because you have an App.java class and thus the resultFileProcessor will always exist. Am I wrong?

Comment (MartinWitt) [reply src/test/java/fr/inria/gforge/spoon/mojo/SpoonMojoTest.java]: from my understanding the test will always fail, because we need the processors from the "/resources/processors/"

Comment (vmassol) [reply src/test/java/fr/inria/gforge/spoon/mojo/SpoonMojoTest.java]: Note: on my side I didn't try the test I wrote because the maven build was failing.

The processor is at https://github.com/SpoonLabs/spoon-maven-plugin/blob/4bd92ae585a0004ec3c587445c71c5dd3f7c5133/src/test/java/fr/inria/gforge/spoon/mojo/CountStatementProcessor.java#L10, ,no? So it should be there in the CP.

What I meant is that you're testing that the check worked by verifying that the output file exists. But this file will be created by the processor running on the main source, no?

Side note: this is why I didn't have any App.java class in my test.

Comment (vmassol) [reply src/test/java/fr/inria/gforge/spoon/mojo/SpoonMojoTest.java]: All this is speculation since I didn't run the maven build :)

**Comment by MartinWitt** [comment]
now it should only test with the test source file.

File: src/test/java/fr/inria/gforge/spoon/mojo/SpoonMojoTest.java (hunk 1 above)
Comment (MartinWitt) [reply src/test/java/fr/inria/gforge/spoon/mojo/SpoonMojoTest.java]: having only test classes is a way better test, good point. Changed it.

**Comment by MartinWitt** [comment]
Can i remove the edit in the original issue and see this as solved/ready for review/merge?

**Comment by vmassol** [comment]
> Can i remove the edit in the original issue and see this as solved/ready for review/merge?

Is that a question for me? :) If so, not sure what you mean.

**Comment by MartinWitt** [comment]
Yes was a question for you because i added 
> Edit: nvm vmassol found something that needs to be fixed sry

 Afterwards, because you found some open problem

**Comment by vmassol** [comment]
@MartinWitt I've closed https://github.com/SpoonLabs/spoon-maven-plugin/pull/68 if that's what you meant. Thanks!

File: src/main/java/fr/inria/gforge/spoon/SpoonMojoGenerate.java [hunk 2]
```diff
@@ -136,6 +139,17 @@
 	)
 	private boolean skipSpoonErrors;
 
+	/**
```
Comment (monperrus) [inline src/main/java/fr/inria/gforge/spoon/SpoonMojoGenerate.java:142]: why is this change required? where is it used?

File: src/main/java/fr/inria/gforge/spoon/SpoonMojoGenerate.java [hunk 3]
```diff
@@ -359,4 +373,11 @@ public boolean getSkipGeneratedSources() {
 	public boolean getSkipSpoonErrors() {
 		return skipSpoonErrors;
 	}
+	public File getTestFolder() {
```
Comment (monperrus) [inline src/main/java/fr/inria/gforge/spoon/SpoonMojoGenerate.java:376]: remove?

File: src/test/projects/hello-world-with-test/pom.xml [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  <groupId>fr.inria.gforge.spoon</groupId>
+  <artifactId>hello-world</artifactId>
+  <version>1.0-SNAPSHOT</version>
+
+  <build>
+    <plugins>
+      <plugin>
+        <groupId>fr.inria.gforge.spoon</groupId>
+        <artifactId>spoon-maven-plugin</artifactId>
+        <configuration>
```
Comment (monperrus) [inline src/test/projects/hello-world-with-test/pom.xml:14]: the processor should be configured in full classpath

File: src/test/projects/hello-world-with-test/test/main/java/fr/inira/gforge/spoon/AppTest.java [hunk 5]
```diff
@@ -0,0 +1,8 @@
+package fr.inria.gforge.spoon;
+
+public class AppTest {
```
Comment (monperrus) [inline src/test/projects/hello-world-with-test/test/main/java/fr/inira/gforge/spoon/AppTest.java:3]: the code to be checked should contain a reference to a test dependency such as junit

File: src/main/java/fr/inria/gforge/spoon/SpoonMojoGenerate.java (hunk 2 above)
Comment (MartinWitt) [reply src/main/java/fr/inria/gforge/spoon/SpoonMojoGenerate.java:142]: it gives you the option to split testFiles and srcFiles in handling, if i understood the code correct. Currently it's not used, but allows future extensions easier or?

File: src/test/projects/hello-world-with-test/pom.xml (hunk 4 above)
Comment (MartinWitt) [reply src/test/projects/hello-world-with-test/pom.xml:14]: I dont understand what you mean sry. Could you explain it?

Comment (vmassol) [reply src/test/projects/hello-world-with-test/pom.xml:14]: > the processor should be configured in full classpath

@monperrus  BTW could you explain why you've removed fullclasspath by default? It seemed like it would have better chance to work by default than having to force a configuration parameter, no? What am I missing? And yes I'm sure there are good reasons I don't know, I haven't followed this discussion/decision. I'm just curious to know the reasons. Thx!

Comment (vmassol) [reply src/test/projects/hello-world-with-test/pom.xml:14]: > I dont understand what you mean sry. Could you explain it?

I think @monperrus means using `<noclasspath>false</noclasspath>` or something like this but I'm not sure either.

File: src/test/projects/hello-world-with-test/test/main/java/fr/inira/gforge/spoon/AppTest.java (hunk 5 above)
Comment (MartinWitt) [reply src/test/projects/hello-world-with-test/test/main/java/fr/inira/gforge/spoon/AppTest.java:6]: added.

File: src/main/java/fr/inria/gforge/spoon/SpoonMojoGenerate.java (hunk 2 above)
Comment (monperrus) [reply src/main/java/fr/inria/gforge/spoon/SpoonMojoGenerate.java:142]: If it's not used, then let's remove it

File: src/test/projects/hello-world-with-test/pom.xml (hunk 4 above)
Comment (monperrus) [reply src/test/projects/hello-world-with-test/pom.xml:14]: > @monperrus BTW could you explain why you've removed fullclasspath by default?

It was removed from spoon-core by default, we did not make any explicit decision for spoon-maven-plugin, it inherits the default of spoon-core.

Why we set the default to NOCLASSPATH in spoon-core? Because it mawimizes the chance that a newcomer gets something interesting out of spoon without reading the doc and without setting a correct full classpath.

**Comment by monperrus** [comment]
Now, we need to expand `.travis.xml to check that spoon-maven-plugin works on `hello-world-with-test/pom.xml` from the command-line. Thanks!

File: src/main/java/fr/inria/gforge/spoon/SpoonMojoGenerate.java (hunk 2 above)
Comment (MartinWitt) [reply src/main/java/fr/inria/gforge/spoon/SpoonMojoGenerate.java]: removed

**Comment by MartinWitt** [comment]
Still need to lookup, how to specify classpath entries, for using <noClasspath>true</noClasspath>. Rest of review should be resolved

**Comment by monperrus** [comment]
Many many thanks @MartinWitt for the progress here.

> Still need to lookup, how to specify classpath entries, for using <noClasspath>false</noClasspath>. 

See https://github.com/SpoonLabs/spoon-maven-plugin/pull/69

@vmassol Does this meet your use case?

**Comment by MartinWitt** [comment]
@monperrus my problem is, how can i tell spoon the classpath entries from the maven plugin?
As you see the CI fails, because the junit classpath entry is missing.
I expected to have something like : `<classpath>$String</classpath>`
(btw: i think the double negation use `noClassPath: false` is pretty complex for a normal user)

**Comment by MartinWitt** [comment]
Okay i found the problem, currently `<scope>test</scope>`  does not work, in classpath mode. looking into it later.
```
      <dependency>
      <groupId>junit</groupId>
      <artifactId>junit</artifactId>
      <version>4.12</version>
      <scope>test</scope>
  </dependency>```

**Comment by monperrus** [comment]
OK, cool.

**Comment by MartinWitt** [comment]
After looking into it, it seems only the fullClasspathMode has still some problems. My random guess the worktime to fix it is a few hours, but a clean solution could take longer. But it seems like @vmassol wants the noclasspathmode fixed and i know the feeling waiting for PRs merge, we could do the following:
- Split the bugreport in 2 parts noClasspath and fullClasspath.
- Only merge the noclasspathmode fix ( remove the fullclasspath part from the CI etc. ).
- Let me afterwards investigate the fullclasspath problem and hope i find a clean solution without rewritting the "ClasspathHacker"
wdyt @monperrus  ?

**Comment by vmassol** [comment]
Thanks @MartinWitt :) 

Is there some doc somewhere to explain what this classpath mode vs noclasspath mode means? I really have no clue what it mean and how to use it at this point. So I don't even know if I need to do something special on our usage of spoon on xwiki.

**Comment by MartinWitt** [comment]
oh sry i missread it. You use the fullclasspath mode. Then the split is useless.
Current problem is, that somehow maven doesn't give me dependencies with `<scope>test</scope>`.
Looking into it later again.

**Comment by vmassol** [comment]
> Current problem is, that somehow maven doesn't give me dependencies with <scope>test</scope>.

I'd have expected https://github.com/SpoonLabs/spoon-maven-plugin/pull/68/files#diff-0b7daf18c571b5ccd8c70d6e9c1d4277R17 to make `project.getArtifacts()`  return all matching. Maybe you need more. There used to be a `project.getTestArtifacts()` which is deprecated, we need to find by what it's been replaced. I thought it was by https://github.com/SpoonLabs/spoon-maven-plugin/pull/68/files#diff-0b7daf18c571b5ccd8c70d6e9c1d4277R17

**Comment by monperrus** [comment]
> In Maven you always have the access to all dependencies (to build a full classpath), so why wouldn't the spoon maven plugin always use it (it seems to me that no using it will always give less good results than using it)?

I think I agree with you but we may miss some corner cases, let's see what the community says on https://github.com/INRIA/spoon/issues/3345

File: src/main/java/fr/inria/gforge/spoon/configuration/AbstractSpoonConfigurationBuilder.java [hunk 6]
```diff
@@ -115,6 +117,18 @@ public SpoonConfigurationBuilder addSourceClasspath() throws SpoonMavenPluginExc
 		List<String> compileClasspath;
 		try {
 			compileClasspath = project.getCompileClasspathElements();
+			String userHome = System.getProperty("user.home");
+			File mavenHome = new File(userHome + "/.m2/repository");
```
Comment (vmassol) [inline src/main/java/fr/inria/gforge/spoon/configuration/AbstractSpoonConfigurationBuilder.java:121]: ouch that's bad.....

Comment (vmassol) [reply src/main/java/fr/inria/gforge/spoon/configuration/AbstractSpoonConfigurationBuilder.java:121]: you shoudn't do that. This is hardcoded, you should leave it to maven.

File: src/main/java/fr/inria/gforge/spoon/configuration/AbstractSpoonConfigurationBuilder.java [hunk 7]
```diff
@@ -115,6 +117,18 @@ public SpoonConfigurationBuilder addSourceClasspath() throws SpoonMavenPluginExc
...
+			for (Dependency dependency : project.getDependencies()) {
+				String groupPath = dependency.getGroupId().replaceAll("\\.", "/");
+				String artifactName = dependency.getArtifactId();
+				String version = dependency.getVersion();
+				File entry = Paths.get(mavenHome.toString(), groupPath, artifactName, version,
+						artifactName + "-" + version + "." + dependency.getType()).toFile();
```
Comment (vmassol) [inline src/main/java/fr/inria/gforge/spoon/configuration/AbstractSpoonConfigurationBuilder.java:127]: This is also bad since this should be left to maven.

Comment (vmassol) [reply src/main/java/fr/inria/gforge/spoon/configuration/AbstractSpoonConfigurationBuilder.java:127]: I really think this is definitely not the way to go :)

Comment (MartinWitt) [reply src/main/java/fr/inria/gforge/spoon/configuration/AbstractSpoonConfigurationBuilder.java:127]: going to write a comment later on it.

**Comment by MartinWitt** [comment]
tl;dr; plugin version was not set, so maven used a old one. Long debug session, "much fun", it's fixed.
@vmassol one last final review? 
sry for the fix try yesterday, looking at it today its real not the way to solve a problem.

File: README.md [hunk 8]
```diff
@@ -159,7 +159,7 @@ Maps and objects are created like this:
 
 ## How to change source and output folder?
 
-You can specify at spoon its input and output directories with, respectively, `srcFolder` and `outFolder` tags.
+You can specify at spoon its input and output directories with, respectively, `srcFolder`, `testFolder` and `outFolder` tags.
```
Comment (monperrus) [inline README.md:162]: I'm note sure this new XML tag is actually supported (It's not in the new pom.xml)

Here, we could say that "By default, spoon-maven-plugin considers the source folders (application sources and test sources) configured in the pom.xml"

Comment (MartinWitt) [reply README.md]: correct, 2 testFolder parameter was part of revision 1 and isn't there anymore. Added your sentence and removed the testFolder tag.

Comment (monperrus) [reply README.md]: Now that we fully rely on the Maven config, I wonder whether we need to keep `srcFolder`. We could directly rely on the Maven `source` configuration. I'm not sure this is for this PR, but I would like to have your opinion on this. WDYT @MartinWitt @vmassol?

Comment (vmassol) [reply README.md]: Yes,TBH I've never understood why the spoon maven plugin was adding this config parameter.  Also it's already possible to add extra src/main or src/test directories using the build helper plugin (https://www.mojohaus.org/build-helper-maven-plugin/index.html) if need be.

Comment (vmassol) [reply README.md]: Note that you still need a config param (boolean) to say whether to run spoon on test sources or not.

Comment (MartinWitt) [reply README.md]: maybe a bit longer than needed and you can skip to conclusion directy but:
### user
From my point of view, the maven plugin has 3 usergroups: starter, programmer and expert.
Start has zero maven knowledge and some java knowledge.
Programmer knows how to use java and can use maven to some degree.
Expert knows java and maven well, to understand this plugin and future configs easily.

### usecases

We can split a java project in 3 folder groups (groups is here an array of paths).
source: main source code
test: test code containg unit test.
other: maybe some resources, used in unit tests or generated code.
For every folder group, we can easily think of a need for checking some policies with spoon.

Spoon wants to provide an API to let all three user use spoon, with all three folder groups in every combination( e.g check only src and other).
So we want an easy default config but allow configuration to most needs.
A starter user shouldn't need to lookup other maven plugins.

Currently a user needs to define the input folders. As said, maven could do this for the user.
A starter could use the plugin faster, if we add some default setting matching this.
But maybe a user wants to check only one package, so he still needs the input path option.
### Conclusion
To match all user needs, i can currently think of we could add two new booleans, deprecate the srcFolder setting and add inputPath, (*or* keep srcFolder even if it sounds missleading).
`<includeSrc>$Boolean</includeSrc>`
`<includeTest>$Boolean</includeTest>`
`<inputPath>$String</inputPath>`

For the includeSrc/includeTest setting we simply use the maven api and it should be correct in almost all cases. The inputPath setting is the old srcFolder setting. As a default setting i would propose:
`<includeSrc>true</includeSrc>`
`<includeTest>false</includeTest>`
`<inputPath></inputPath>`
  
#### is a boolean flag for includeTest sufficent
No, because what do you do if you want to run analysis on test and not source?

wdyt?

**Comment by monperrus** [comment]
We're on the same line now. To recap:

* remove `srcFolder`
  * Issue: https://github.com/INRIA/spoon/issues/3345
  * PR: none
* <includeSrc>true</includeSrc>
  * Issue: https://github.com/INRIA/spoon/issues/3349
  * PR: none
* `<includeTest>false</includeTest>`
  * Issue: https://github.com/INRIA/spoon/issues/3326
  * PR: https://github.com/SpoonLabs/spoon-maven-plugin/pull/68 (this one)

We can address them in three different PRs.

> is a boolean flag for includeTest sufficient?

For https://github.com/INRIA/spoon/issues/3326 and this PR, I'd say yes

**Comment by MartinWitt** [comment]
Would you drastically remove srcFolder instantly or deprecate it with a removal date in X months?
Deprecated things have the bad habit of living forever and instant removal could annoy users.
What's your preference?  Personally i hate deprecated things, because it generates technical debt and is resolved in most cases, when the project doesn't build anymore.

**Comment by monperrus** [comment]
> Would you drastically remove srcFolder instantly?


Yes. This is a very rarely used feature to my knowledge.

In Spoon, we make an informed decision each time we need to remove something, if it's highly used we
deprecate until the next release, otherwise we remove.

Thanks!

File: src/main/java/fr/inria/gforge/spoon/configuration/AbstractSpoonConfigurationBuilder.java [hunk 9]
```diff
@@ -36,35 +36,19 @@ protected AbstractSpoonConfigurationBuilder(SpoonMojoGenerate spoon,
...
-
-		String inputs = "";
-		for (int i = 0; i < srcDir.size(); i++) {
-			File file = srcDir.get(i);
-			inputs += file.getAbsolutePath();
-			if (i != srcDir.size() - 1) {
-				inputs += File.pathSeparatorChar;
-			}
-		}
+		String inputs = srcDir.stream().map(File::getAbsolutePath).collect(Collectors.joining(File.pathSeparator));
```
Comment (MartinWitt) [inline src/main/java/fr/inria/gforge/spoon/configuration/AbstractSpoonConfigurationBuilder.java:51]: does the same but is much more readable. wdyt?

File: src/main/java/fr/inria/gforge/spoon/configuration/AbstractSpoonConfigurationBuilder.java [hunk 10]
```diff
@@ -36,35 +36,19 @@ protected AbstractSpoonConfigurationBuilder(SpoonMojoGenerate spoon,
...
-				srcDir.add(new File(spoon.getProject().getBuild().getSourceDirectory()));
-			}
+		if(spoon.isIncludeSrcDirectories()) {
+			srcDir.add(new File(spoon.getProject().getBuild().getSourceDirectory()));
 		}
-
+		if(spoon.isIncludeTestDirectories()) {
+			srcDir.add(new File(spoon.getProject().getBuild().getTestSourceDirectory()));
+		}
+		//TODO: Fallback with generated resources needed? Maven should always know src and test folder way better than us or?
```
Comment (MartinWitt) [inline src/main/java/fr/inria/gforge/spoon/configuration/AbstractSpoonConfigurationBuilder.java:45]: is the old fallback still needed?

File: src/test/java/fr/inria/gforge/spoon/mojo/SpoonMojoTest.java [hunk 11]
```diff
@@ -141,7 +139,7 @@ public void testSpoonThrowException() throws Exception {
 		final File dirOutputResults = new File(basedir, "target/spoon-maven-plugin");
 		assertThat(dirOutputResults).doesNotExist();
 	}
-
+	@Ignore("source folder could point to a non existent folder early, asking maven for source folders removes this.")
```
Comment (MartinWitt) [inline src/test/java/fr/inria/gforge/spoon/mojo/SpoonMojoTest.java:142]: error is "lost" because see comment. remove the testcase?

File: src/test/projects/custom-configuration/pom.xml [hunk 12]
```diff
@@ -8,11 +8,29 @@
...
+              <goal>add-source</goal>
+            </goals>
+            <configuration>
+              <sources>
+                <source>src/internal/fr/inria/gforge/spoon</source>
+              </sources>
+            </configuration>
+          </execution>
+        </executions>
+      </plugin>
```
Comment (MartinWitt) [inline src/test/projects/custom-configuration/pom.xml:29]: Add to readme, under "How to add more input folders"?

File: README.md [hunk 13]
```diff
@@ -160,6 +160,7 @@ Maps and objects are created like this:
 ## How to change source and output folder?
 
 You can specify at spoon its input and output directories with, respectively, `srcFolder` and `outFolder` tags.
```
Comment (MartinWitt) [inline README.md:162]: still needs rewrite.

**Comment by MartinWitt** [comment]
The testConfigurationOfTheCustomInputFolder shows, that removing <srcFolder> option removes the ability to check only some custom folder. Readd the sourceFolder setting or lose the ability?

File: src/main/java/fr/inria/gforge/spoon/configuration/AbstractSpoonConfigurationBuilder.java (hunk 9 above)
Comment (monperrus) [reply src/main/java/fr/inria/gforge/spoon/configuration/AbstractSpoonConfigurationBuilder.java:51]: good!

File: src/main/java/fr/inria/gforge/spoon/configuration/AbstractSpoonConfigurationBuilder.java (hunk 10 above)
Comment (monperrus) [reply src/main/java/fr/inria/gforge/spoon/configuration/AbstractSpoonConfigurationBuilder.java:45]: no.

File: src/test/java/fr/inria/gforge/spoon/mojo/SpoonMojoTest.java (hunk 11 above)
Comment (monperrus) [reply src/test/java/fr/inria/gforge/spoon/mojo/SpoonMojoTest.java:142]: I'd say yes

File: README.md (hunk 13 above)
Comment (monperrus) [reply README.md:162]: ok

**Comment by monperrus** [comment]
OK for me to remove this feature.

**Comment by MartinWitt** [comment]
All todos are resolved now.

**Comment by monperrus** [comment]
Thanks a lot @MartinWitt , that was long but very important work to be done.

**Comment by monperrus** [comment]
Thanks @vmassol for all your inputs.

**Comment by vmassol** [comment]
@monperrus @MartinWitt Any idea when a spoon version containing this will be available? :) Thanks!

**Comment by MartinWitt** [comment]
Sadly releases are out of my scope, you have to wait for a maintainer(?) for a release. 
But i see no reason to not release a new version 3.4 after this breaking change.

**Comment by monperrus** [comment]
I'll do the release.

@vmassol do you confirm that everything works as expected on your setup?

**Comment by vmassol** [comment]
bq. @vmassol do you confirm that everything works as expected on your setup?

I was actually waiting for the release to test :) Problem is that I probably won't have the time today to test it (I'm in meeting all day). Now the issue represents my need for sure and you have tests for it I guess so it should be fine. If you prefer me to test it, then I could do that Friday.

Thx

**Comment by monperrus** [comment]
> If you prefer me to test it, then I could do that Friday.

Yes, that would be better. We're not in a rush anyway.

**Comment by vmassol** [comment]
> Yes, that would be better. We're not in a rush anyway.

Actually I won't have the time because of the release. Maybe next week, but not sure yet.

The "rush" was because I wanted to have this before the 12.4RC1 release of XWiki but from what I see it's going to be too late (I don't like to have an issue with commits and not closed for a release). But it's ok, it'll go in the next release in one month. It just means the spoon check is not active on xwiki FTM so more violations can be added any day. Definitely not critical but nice to have ASAP.

My point is that you should not wait for me. You need to have your own tests so that you're sure that it works.

But I'll be happy to help if I can, but will take some time.

Thx

**Comment by monperrus** [comment]
No worries, we move forward.

Release 3.4 done! 

@MartinWitt would you prepare a changelog to be added to https://github.com/SpoonLabs/spoon-maven-plugin/releases/tag/spoon-maven-plugin-3.4?

Thanks!

**Comment by vmassol** [comment]
Thanks @monperrus!

**Comment by vmassol** [comment]
@MartinWitt  I think this comment needs to be modified since you've removed the testFolder config properties: https://github.com/SpoonLabs/spoon-maven-plugin/blob/4f6de46bf751446248dcb3db66f816a8c15ce241/src/main/java/fr/inria/gforge/spoon/SpoonMojoGenerate.java#L37

Thanks

**Comment by vmassol** [comment]
@MartinWitt @monperrus I think I've found a regression in v3.4, see https://github.com/INRIA/spoon/issues/3366

**Comment by MartinWitt** [comment]
@monperrus changelog comes after fixing the regression if that is okay for you?
@vmassol looking into the comment, after fixing the new bug, if that is okay for you?

**Comment by vmassol** [comment]
Sure, perfectly fine. Thx a lot.

**Comment by monperrus** [comment]
> @monperrus changelog comes after fixing the regression if that is okay for you?

yes! we will release 3.4.1 and changelog it directly
